src/main/java/org/opensearch/flowframework/workflow/Workflow.java

```diff
@@ -0,0 +1,165 @@
+
+        xContentBuilder.startObject(USER_PARAMS_FIELD);
+        for (Entry<String, String> e : userParams.entrySet()) {
+            xContentBuilder.field(e.getKey(), e.getValue());
+        }
+        xContentBuilder.endObject();
+
+        xContentBuilder.startArray(NODES_FIELD);
+        for (WorkflowNode n : nodes) {
+            xContentBuilder.value(n);
+        }
+        xContentBuilder.endArray();
+
+        xContentBuilder.startArray(EDGES_FIELD);
+        for (WorkflowEdge e : edges) {
+            xContentBuilder.value(e);
+        }
+        xContentBuilder.endArray();
+
+        return xContentBuilder.endObject();
+    }
+
+    /**
+     * Parse raw JSON content into a workflow instance.
+     *
+     * @param parser JSON based content parser
+     * @return the parsed Workflow instance
+     * @throws IOException if content can't be parsed correctly
+     */
+    public static Workflow parse(XContentParser parser) throws IOException {
+        Map<String, String> userParams = new HashMap<>();
+        List<WorkflowNode> nodes = new ArrayList<>();
+        List<WorkflowEdge> edges = new ArrayList<>();
+
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            String fieldName = parser.currentName();
+            parser.nextToken();
+            switch (fieldName) {
+                case USER_PARAMS_FIELD:
+                    ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        String userParamFieldName = parser.currentName();
+                        parser.nextToken();
+                        userParams.put(userParamFieldName, parser.text());
+                    }
+                    break;
+                case NODES_FIELD:
+                case STEPS_FIELD:
+                    ensureExpectedToken(XContentParser.Token.START_ARRAY, parser.currentToken(), parser);
+                    while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                        nodes.add(WorkflowNode.parse(parser));
+                    }
+                    break;
+                case EDGES_FIELD:
+                    ensureExpectedToken(XContentParser.Token.START_ARRAY, parser.currentToken(), parser);
+                    while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                        edges.add(WorkflowEdge.parse(parser));
+                    }
+                    break;
+            }
+
+        }
+        if (nodes.isEmpty()) {
+            throw new IOException("A workflow must have at least one node.");
+        }
+        if (edges.isEmpty()) {
+            // infer edges from sequence of nodes
+            // Start iteration at 1, will skip for a one-node array
+            for (int i = 1; i < nodes.size(); i++) {
```

Comment:
Style/Nit: How about starting from `0` and then doing `i` and `i+1` respectively below?

Reply:
Starting at 1 avoids adding a `-1` to `nodes.size()` which, with the spacing added by spotless, is 4 more bytes of space in the source code. Storage space is expensive! Back in 1960 a byte cost nearly $1!  It's cheaper now at $0.000076 per byte.

They're equivalent, and I prefer this one as I'm really iterating pairs and whether I pick the first or second of the pair isn't that important.  Happy to switch, though, I expect they compile to the same byte code.